Introduce Span input and output support (#33)
iCharlesHu/swift-experimental-subprocess: README.md

## Proposed solution

We propose a new package, `Subprocess`, that will eventually replace `Process` as the canonical way to launch a process in `Swift`.

Here's the above script rewritten using `Subprocess` package:

```swift
#!/usr/bin/swift

import Subprocess

> let gitResult = try await Subprocess.run(   // <- 0

Reviewer: ```suggestion
let gitResult = try await run(   // <- 0
```

Author: On in testing I found you can't actually omit the `Subprocess` here otherwise you get
```
Use of 'run' refers to instance method rather than global function 'run(_:arguments:environment:workingDirectory:platformOptions:input:output:error:)' in module 'Subprocess'
```

Author: ^^ That only existed under `XCTestCase` because of the name `run()`.

/// - Returns a ExecutableResult type containing the return value
///     of the closure.
public func run<Result>(
    _ executable: Executable,
    arguments: Arguments = [],
    environment: Environment = .inherit,
    workingDirectory: FilePath? = nil,
    platformOptions: PlatformOptions = PlatformOptions(),
    isolation: isolated (any Actor)? = #isolation,
    body: (@escaping (Execution<SequenceOutput, DiscardedOutput>, StandardInputWriter) async throws -> Result)
) async throws -> ExecutionResult<Result>

// MARK: - Configuration Based
/// Run a executable with given parameters specified by a `Configuration`
/// - Parameters:
///   - configuration: The `Subprocess` configuration to run.
///   - output: The method to use for redirecting the standard output.
///   - error: The method to use for redirecting the standard error.
///   - body: The custom configuration body to manually control
///       the running process and write to its standard input.
/// - Returns a ExecutableResult type containing the return value
///     of the closure.
public func run<Result, Output: OutputProtocol, Error: OutputProtocol>(
    _ configuration: Configuration,

>     isolation: isolated (any Actor)? = #isolation,

Reviewer 2: CAn we make sure the order of the arguments stay consistent? Previously the arguments are in this order: 
- input
- output
- error
- isolation
- body

But here isolation is moved up to before output and error. Same with the one below

Author: Ahh good catch! Thanks!

## Detailed Design

### The `run()` Family of Methods

> We propose several `run()` functions that allows developers to asynchronously executes a subprocess.

Reviewer: ```suggestion
We propose several `run()` functions that allow developers to asynchronously execute a subprocess.
```

```swift
/// Run a executable with given parameters and a custom closure
/// to manage the running subprocess' lifetime and its IOs.
/// - Parameters:
///   - executable: The executable to run.
///   - arguments: The arguments to pass to the executable.
///   - environment: The environment in which to run the executable.
///   - workingDirectory: The working directory in which to run the executable.
///   - platformOptions: The platform specific options to use
///     when running the executable.
///   - input: The input to send to the executable.
///   - output: The method to use for redirecting the standard output.
///   - error: The method to use for redirecting the standard error.
///   - body: The custom execution body to manually control the running process
/// - Returns a CollectedResult containing the result of the run.

> @available(macOS 9999, *)

Reviewer: What kind of availability do we really need here? I assume it's 9999 because we don't have a macOS with `Span` yet?

Author: Remove unnecessary ones and just comment on the proposal that's it's Span.

///   - configuration: The `Subprocess` configuration to run.
///   - input: A file descriptor to bind to the subprocess' standard input.
///   - output: A file descriptor to bind to the subprocess' standard output.
///   - error: A file descriptor to bind to the subprocess' standard error.
/// - Returns: the process identifier for the subprocess.
@available(macOS 9999, *)
public func runDetached(
    _ configuration: Configuration,
    input: FileDescriptor? = nil,
    output: FileDescriptor? = nil,
    error: FileDescriptor? = nil
) throws -> ProcessIdentifier
```

### `Execution`

In contrast to the monolithic `Process`, `Subprocess` utilizes two types to model the lifetime of a process. `Configuration` (discussed later) and `Execution`. `Execution` is designed to represent an executed process. This execution could be either in progress or completed. Direct construction of `Execution` instances is not supported; instead, a `Execution` object is passed to the `body` closure of `run()`. This object is only valid within the scope of the closure, and developers may use it to send signals to the child process or retrieve the child's standard I/Os via `AsyncSequence`s.

```swift

> /// An object that repersents a subprocess that has been

Reviewer: ```suggestion
/// An object that represents a subprocess that has been
```

/// executed. You can use this object to send signals to the
/// child process as well as stream its output and error.
@available(macOS 9999, *)

> public struct Execution<

Reviewer: If this is only valid in the scope of the closure, does it make sense to mark it as `~Escapable`?

    Output: OutputProtocol,
    Error: OutputProtocol
>: Sendable {
    /// The process identifier of the current execution
    public let processIdentifier: ProcessIdentifier
}

@available(macOS 9999, *)
extension Execution where Output == SequenceOutput {
    /// The standard output of the subprocess.
    /// Accessing this property will **fatalError** if
    /// - `.output` wasn't set to `.redirectToSequence` when the subprocess was spawned;
    /// - This property was accessed multiple times. Subprocess communicates with

>     ///   parent process via pipe under the hood and each pipe can only be consumed ones.

Reviewer: ```suggestion
    ///   parent process via pipe under the hood and each pipe can only be consumed once.
```
(this typo is in a couple of places)

    public var standardOutput: some AsyncSequence<Data, any Swift.Error>
}

@available(macOS 9999, *)
extension Execution where Error == SequenceOutput {
    /// The standard error of the subprocess.
    /// Accessing this property will **fatalError** if
    /// - `.error` wasn't set to `.redirectToSequence` when the subprocess was spawned;
    /// - This property was accessed multiple times. Subprocess communicates with
    ///   parent process via pipe under the hood and each pipe can only be consumed ones.
    public var standardError: some AsyncSequence<Data, any Swift.Error>
}

#if canImport(WinSDK)
/// A platform independent identifier for a subprocess.

> @available(macOS 9999, *)

Reviewer: What does this availability mean for Windows?

```swift
#if canImport(Glibc) || canImport(Darwin)
/// A step in the graceful shutdown teardown sequence.
/// It consists of a signal to send to the child process and the
/// duration allowed for the child process to exit before proceeding
/// to the next step.

> @available(FoundationPreview 0.4, *)

Reviewer: I think this availability should be updated to match the rest, whatever we decide.

@available(iOS, unavailable)
@available(tvOS, unavailable)
@available(watchOS, unavailable)

> public struct TeardownStep: Sendable, Hashable {

Reviewer: It's not immediately clear to me what the name of this type is really meant for. It's not possible to instantiate it, right? Should this function be static on `Signal` instead?

```

### `Configuration`

`Configuration` represents the collection of information needed to spawn a process. This type is designed to be very similar to the existing `Process`, enabling you to configure your process in a manner akin to `NSTask`:

```swift
/// A collection of configurations parameters to use when
/// spawning a subprocess.
@available(FoundationPreview 0.4, *)
@available(iOS, unavailable)
@available(tvOS, unavailable)
@available(watchOS, unavailable)
public struct Configuration : Sendable, Hashable {
    /// The executable to run.
    public var executable: Executable
    /// The arguments to pass to the executable.
    public var arguments: Arguments
    /// The environment to use when running the executable.
    public var environment: Environment
    /// The working directory to use when running the executable.
    public var workingDirectory: FilePath
    /// The platform specifc options to use when
    /// running the subprocess.
    public var platformOptions: PlatformOptions

    public init(
        executing executable: Executable,
        arguments: Arguments = [],
        environment: Environment = .inherit,

>         workingDirectory: FilePath? = nil,

Reviewer: Should the ivar also be `FilePath?` instead of non-nullable?

        platformOptions: PlatformOptions = .default
    )
}

extension Configuration : CustomStringConvertible, CustomDebugStringConvertible {}
```

**Note:** the `.workingDirectory` property defaults to the current working directory of the calling process.

### `StandardInputWriter`

`StandardInputWriter` provides developers with direct control over writing to the child process's standard input. Similar to the `Execution` object, developers should use the `StandardInputWriter` object passed to the `body` closure, and this object is only valid within the body of the closure.

**Note**: Developers should call `finish()` when they have completed writing to signal that the standard input file descriptor should be closed.

```swift
/// A writer that writes to the standard input of the subprocess.
@available(FoundationPreview 0.4, *)
@available(iOS, unavailable)
@available(tvOS, unavailable)
@available(watchOS, unavailable)
public final actor StandardInputWriter {
    /// Write a sequence of UInt8 to the standard input of the subprocess.
    /// - Parameter sequence: The sequence of bytes to write.

>     public func write<SendableSequence: Sequence<UInt8> & Sendable>(

Reviewer: I'm still unsure about this - do we really want to support *any* sequence of `UInt8` or do we want to require a `Collection` instead so we know we can do a bulk operation? Even if we want to allow it, shouldn't we provide concrete overrides of this that take a `Collection` or some other known-contiguous types like `[UInt8]` to avoid the overhead of casting in the implementation?

Author: Explicit `ArrayInput` and `DataInput`.

        _ sequence: SendableSequence
    ) async throws
    /// Write a String to the standard input of the subprocess.
    /// - Parameter sequence: The string write.

>     public func write(

Reviewer: Does this deserve a defaulted-to-`.utf8` encoding argument?

    // Set list of supplementary group IDs for the subprocess
    public var supplementaryGroups: [gid_t]?
    /// Set the process group for the subprocess, equivalent to
    /// calling `setpgid()` on the child process.
    /// Process group ID is used to group related processes for
    /// controlling signals.
    public var processGroupID: pid_t?
    // Creates a session and sets the process group ID
    // i.e. Detach from the terminal.
    public var createSession: Bool
    // Whether the subprocess should close all file
    // descriptors except for the ones explicitly passed
    // as `input`, `output`, or `error` when `run` is executed
    // This is equivelent to setting `POSIX_SPAWN_CLOEXEC_DEFAULT`
    // on Darwin. This property is default to be `false`
    // because `POSIX_SPAWN_CLOEXEC_DEFAULT` is a darwin-specific
    // extension and we can only emulate it on Linux.
    public var closeAllUnknownFileDescriptors: Bool
    /// An ordered list of steps in order to tear down the child
    /// process in case the parent task is cancelled before
    /// the child proces terminates.
    /// Always ends in sending a `.kill` signal at the end.

>     public var teardownSequence: [TeardownStep] = []

Reviewer: I guess this is where these are used, but how do you create them other than the one static func?

#endif // canImport(Glibc)
```

Similar to the Darwin version, the Linux `PlatformOptions` also has an "escape hatch" closure that allows the developers to explicitly configure the subprocess. This closure is run after `fork` but before `exec`. In the example below, `preSpawnProcessConfigurator` can be used to set the group ID for the subprocess:

```swift
var platformOptions: PlatformOptions = .default
// Set Group ID for process
platformOptions.preSpawnProcessConfigurator = {
    setgid(4321)
}
```

### `PlatformOptions` on Windows

On Windows, we propose the following `PlatformOptions`:

```swift
#if canImport(WinSDK)
/// The collection of platform-specific settings
/// to configure the subprocess when running
@available(FoundationPreview 0.4, *)
@available(iOS, unavailable)
@available(tvOS, unavailable)
@available(watchOS, unavailable)
public struct PlatformOptions: Sendable, Hashable {
    public struct UserCredentials: Sendable, Hashable {
        // The name of the user. This is the name
        // of the user account to run as.
        public var username: String
        // The clear-text password for the account.

>         public var password: String

Reviewer: I assume this mirrors an underlying Win32 API?

Author: Yes. As bad as it is the Win32 API literately says "plain text password"... 😅

- `StringOutput`: collects output from the subprocess as `String` with the given encoding.

```swift
/// `OutputProtocol` specifies the set of methods that a type
/// must implement to serve as the output target for a subprocess.
/// Instead of developing custom implementations of `OutputProtocol`,
/// it is recommended to utilize the default implementations provided
/// by the `Subprocess` library to specify the output handling requirements.
public protocol OutputProtocol: Sendable {
    associatedtype OutputType: Sendable
    /// Lazily create and return the FileDescriptor for reading
    func readFileDescriptor() throws -> FileDescriptor?
    /// Lazily create and return the FileDescriptor for writing
    func writeFileDescriptor() throws -> FileDescriptor?
    /// Return the read `FileDescriptor` and remove it from the output
    /// such that the next call to `consumeReadFileDescriptor` will
    /// return `nil`.
    func consumeReadFileDescriptor() -> FileDescriptor?

    /// Close the FileDescriptor for reading
    func closeReadFileDescriptor() throws
    /// Close the FileDescriptor for writing
    func closeWriteFileDescriptor() throws

>     /// Capture the output from the subprocess up to maxSize

Reviewer 2: what is maxSize?

}

/// A concrete `Input` type for subprocesses that indicates
/// the absence of input to the subprocess. On Unix-like systems,
/// `NoInput` redirects the standard input of the subprocess
/// to `/dev/null`, while on Windows, it does not bind any
/// file handle to the subprocess standard input handle.
public struct NoInput: InputProtocol { }

/// A concrete `Input` type for subprocesses that
/// reads input from a specified `FileDescriptor`.
/// Developers have the option to instruct the `Subprocess` to
/// automatically close the provided `FileDescriptor`
/// after the subprocess is spawned.
public struct FileDescriptorInput: InputProtocol { }

/// A concrete `Input` type for subprocesses that reads input
/// from a given type conforming to `StringProtocol`.
/// Developers can specify the string encoding to use when
/// encoding the string to data, which defaults to UTF-8.
public struct StringInput<
    InputString: StringProtocol & Sendable
>: PipeBasedInputProtocol { }

/// A concrete `Input` type for subprocesses that reads input
/// from a given sequence of `UInt8` (such as `Data` or `[UInt8]`).

> public struct UInt8SequenceInput<

Reviewer: I have the same question here as above - do we really want to provide this? Because accidentally falling back to a byte-by-byte write for each element of some collection could end up being very expensive, and surprising.

        _ sequence: InputSequence
    ) -> Self where Self == DataSequenceInput<InputSequence>

    /// Create a Subprocess input from a `AsyncSequence` of `Data`.
    public static func asyncSequence<InputSequence: AsyncSequence & Sendable>(
        _ asyncSequence: InputSequence
    ) -> Self where Self == DataAsyncSequenceInput<InputSequence>

    /// Create a Subprocess input from a type that conforms to `StringProtocol`
    public static func string<InputString: StringProtocol & Sendable>(
        _ string: InputString,
        using encoding: String.Encoding = .utf8
    ) -> Self where Self == StringInput<InputString>
}
```

Here are some examples:

```swift
// By default `InputMethod` is set to `.none`
let ls = try await Subprocess.run(.named("ls"))

// Alternatively, developers could pass in a file descriptor
let fd: FileDescriptor = ...
let cat = try await Subprocess.run(.named("cat"), input: .fileDescriptor(fd, closeAfterSpawningProcess: true))

// Pass in a async sequence directly
let sequence: AsyncSequence<Data> = ...
let exe = try await Subprocess.run(.at("/some/executable"), input: .asyncSequence(sequence))
```

### `OutputProtocol`

> `OutputProtocol` defines the set of methods that a type must implement to serve as the output target for a subprocess. Similarly to `InputProtocol`, developers should utilize the built-in concrete `Output` types provided with `Subprocess` whenever feasible; alternatively, they can create a type that conforms to `PipeBasedOutputProtocol` and implements `func output(from:) throws -> Data` for custom outputs.

Reviewer: ```suggestion
`OutputProtocol` defines the set of methods that a type must implement to serve as the output target for a subprocess. Similarly to `InputProtocol`, developers should utilize the built-in concrete `Output` types provided with `Subprocess` whenever feasible; alternatively, they can create a type that conforms to `PipeBasedOutputProtocol` and implements `func output(from:) throws -> OutputType` for custom outputs.
```

@available(iOS, unavailable)
@available(tvOS, unavailable)
@available(watchOS, unavailable)
extension OutputProtocol where Self == StringOutput {
    /// Create a `Subprocess` output that collects output as
    /// UTF8 String with 128kb limit.
    public static var string: Self {
        return .string(limit: 128 * 1024)
    }

    /// Create a `Subprocess` output that collects output as
    /// `String` using the given encoding.
    public static func string(

>         limit: Int,

Reviewer: It would be good to specify that this value is measured in bytes.

    public let standardError: Error.OutputType
}

@available(macOS 9999, *)
extension CollectedResult: Equatable where Output.OutputType: Equatable, Error.OutputType: Equatable {}

@available(macOS 9999, *)
extension CollectedResult: Hashable where Output.OutputType: Hashable, Error.OutputType: Hashable {}

@available(macOS 9999, *)
extension CollectedResult: Codable where Output.OutputType: Codable, Error.OutputType: Codable {}

@available(macOS 9999, *)
extension CollectedResult: CustomStringConvertible where Output.OutputType: CustomStringConvertible, Error.OutputType: CustomStringConvertible
```

`ExecutionResult` is a simple wrapper around the generic result returned by the `run` closures with the corresponding `TerminationStatus` of the child process:

```swift
/// A simple wrapper around the generic result returned by the
/// `run` closures with the corresponding `TerminationStatus`
/// of the child process.
@available(macOS 9999, *)
public struct ExecutionResult<Result> {
    /// The termination status of the child process
    public let terminationStatus: TerminationStatus
    /// The result returned by the closure passed to `.run` methods
    public let value: Result

>     internal init(terminationStatus: TerminationStatus, value: Result) {

Reviewer: Is this public, or should we omit from the pitch?

    public static func named(_ executableName: String) -> Self
    /// Locate the executable by its full path.
    /// `Subprocess` will use this  path directly.
    public static func at(_ filePath: FilePath) -> Self
    /// Returns the full executable path given the environment value.
    public func resolveExecutablePath(in environment: Environment) -> FilePath?
}

extension Executable : CustomStringConvertible, CustomDebugStringConvertible {}
```

### `Environment`

`struct Environment` is used to configure how should the process being launched receive its environment values:

```swift
@available(FoundationPreview 0.4, *)
@available(iOS, unavailable)
@available(tvOS, unavailable)
@available(watchOS, unavailable)
public struct Environment: Sendable, Hashable {
    /// Child process should inherit the same environment
    /// values from its parent process.
    public static var inherit: Self { get }
    /// Override the provided `newValue` in the existing `Environment`
    public func updating(
        _ newValue: [String : String]
    ) -> Self
    /// Use custom environment variables
    public static func custom(
        _ newValue: [String : String]
    ) -> Self

> #if !os(Windows)

Reviewer: Is Win32 incompatible with `Data`, requiring `String` values only?
